File: go/bond.py
#! /usr/bin/env python3
# [[file:~/Workspace/Programming/chem-utils/chem-utils.note::f48ba944-02bd-4ac0-bc39-6713aa91f65f][f48ba944-02bd-4ac0-bc39-6713aa91f65f]]
from enum import Enum
from collections import namedtuple

from .atom import Atom

class BondOrder(Enum):
    """https://en.wikipedia.org/wiki/Bond_order"""
    disconnected =  0.0
    partial      =  0.5
    single       =  1.0
    aromatic     =  1.5
    double       =  2.0
    tripple      =  3.0
    quadruple    =  4.0

class _Bond(namedtuple("Bond", "atom1 atom2 order".split())):
    """basic bond definition"""
    __slots__ = ()

    def __str__(self):
        return "{}-{}: {}".format(self.atom1.name, self.atom2.name, self.order.name)

    def __eq__(self, other):
        if self.atom1 == other.atom1 and self.atom2 == other.atom2:
            return True
        if self.atom1 == other.atom2 and self.atom2 == other.atom1:
            return True
        return False

# make simple thing simple
def Bond(atom1, atom2, order=BondOrder.single):
    return _Bond(atom1, atom2, BondOrder(order))
# f48ba944-02bd-4ac0-bc39-6713aa91f65f ends here

File: lib/go/src/bond.rs
// [[file:~/Workspace/Programming/chem-utils/chem-utils.note::6d7f16c6-a551-425d-8cbe-e02585ef91f5][6d7f16c6-a551-425d-8cbe-e02585ef91f5]]
use std::fmt;
use std::u64;

#[derive(Clone, Debug)]
pub enum BondType {
    NotConnected,
    Partial,
    Single,
    Aromatic,
    Double,
    Triple,
    Quadruple,
}

#[derive(Clone, Debug)]
pub struct Bond {
    order: f64,
    kind : BondType,
    length: f64,                // for caching
    name: String,
}

impl Bond {
    pub fn new() -> Self{
        Bond {
            order: 1.0,
            kind: BondType::Single,
            length: 0.0,
            name: String::new(),
        }
    }
}

#[test]
fn test_bond() {
    let mut b = Bond::new();
    b.order = 1.2;
    b.name = "c1-c2".to_string();
    b.length = 2.5;
    println!("{:?}", b);
}
// 6d7f16c6-a551-425d-8cbe-e02585ef91f5 ends here
